lightgbm/src/main/scala/com/microsoft/ml/spark/lightgbm/TaskTrainingMethods.scala

```diff
@@ -0,0 +1,164 @@
+            networkInit(nodes, localListenPort, log, LightGBMConstants.NetworkRetries, LightGBMConstants.InitialDelay)
+            if (useSingleDatasetMode) sharedState.doneSignal.await()
+            translate(batchIndex, columnParams, aggregatedValidationColumns, log, trainParams, returnBooster,
+              sharedState.isSparse.get, schema, aggregatedColumns)
+          } else {
+            log.info("Helper task finished processing rows")
+            sharedState.doneSignal.countDown()
+            List[LightGBMBooster]().toIterator
+          }
+        } finally {
+          // Finalize network when done
+          if (isEnabledWorker) LightGBMUtils.validate(lightgbmlib.LGBM_NetworkFree(), "Finalize network")
+        }
+      }
+    } finally {
+      resetState(useSingleDatasetMode, isEnabledWorker, log, sharedState)
+    }
+  }
+
+  def translate(batchIndex: Int, columnParams: ColumnParams, validationData: Option[AggregatedColumns],
+                log: Logger, trainParams: TrainParams, returnBooster: Boolean, isSparse: Boolean, schema: StructType,
+                aggregatedColumns: AggregatedColumns): Iterator[LightGBMBooster] = {
+    var trainDatasetOpt: Option[LightGBMDataset] = None
+    var validDatasetOpt: Option[LightGBMDataset] = None
+    try {
+      beforeGenerateTrainDataset(batchIndex, columnParams, schema, log, trainParams)
+      trainDatasetOpt = DatasetUtils.generateDataset(aggregatedColumns, columnParams, None, isSparse, schema,
+        log, trainParams)
+      afterGenerateTrainDataset(batchIndex, columnParams, schema, log, trainParams)
+
+      if (validationData.isDefined) {
+        beforeGenerateValidDataset(batchIndex, columnParams, schema, log, trainParams)
+        validDatasetOpt = DatasetUtils.generateDataset(validationData.get, columnParams,
+          trainDatasetOpt, isSparse, schema, log, trainParams)
+        afterGenerateValidDataset(batchIndex, columnParams, schema, log, trainParams)
+      }
+
+      var boosterOpt: Option[LightGBMBooster] = None
+      try {
+        val booster = createBooster(trainParams, trainDatasetOpt.get, validDatasetOpt)
+        boosterOpt = Some(booster)
+        val bestIterResult = trainCore(batchIndex, trainParams, booster, log, validDatasetOpt.isDefined)
+        if (returnBooster) {
+          val model = booster.saveToString()
+          val modelBooster = new LightGBMBooster(model)
+          // Set best iteration on booster if hit early stopping criteria in trainCore
+          bestIterResult.foreach(modelBooster.setBestIteration(_))
+          Iterator.single(modelBooster)
+        } else {
+          Iterator.empty
+        }
+      } finally {
+        // Free booster
+        boosterOpt.foreach(_.freeNativeMemory())
+      }
+    } finally {
+      // Free datasets
+      trainDatasetOpt.foreach(_.close())
+      validDatasetOpt.foreach(_.close())
+    }
+  }
+
+  def resetState(useSingleDatasetMode: Boolean, isEnabledWorker: Boolean, log: Logger,
```

Comment:
I think you could remove this if its properly abstracted to create the shared singleton within the fit call

Reply:
let me try this next

Reply:
removed! I tried it out and it seems to be fine when run multiple times (which was the original reason I needed to add this reset state).